### ISSUE TITLE
ListView items color

### ISSUE THREAD

**Issue by aurelienverbeke** [migueldeicaza/gui.cs#1577]
Is there a way to change background/text colors of each item of a ListView ?

**Comment by BDisp** [comment]
Not for the time being. But this may be possible to override rendering and use different color schemes for pairs and odd items.

**Comment by tznind** [comment]
The easiest way is to implement `IListDataSource` (see below).

![image](https://user-images.githubusercontent.com/31306100/150794507-ba3cabd6-a561-4424-8795-2106a2d914e2.png)


```csharp
using System.Collections;
using Terminal.Gui;


Application.Init();

var fancyList = new FancyListRendering();
fancyList.Add("Fish");
fancyList.Add("Dog");
fancyList.Add("Apple");
fancyList.Add("Halibut");
fancyList.Add("Ocean Liner");

var listView = new ListView((IListDataSource)fancyList)
{
	Width = Dim.Fill(),
	Height = Dim.Fill(),
};
var w = new Window()
{
	Width = 20,
	Height = 20
};
w.Add(listView);
Application.Run(w);
Application.Shutdown();


public class FancyListRendering : List<string>, IListDataSource
{
	public int Length => Count;

	public bool IsMarked(int item)
	{
		return false;
	}

	public void Render(ListView container, ConsoleDriver driver, bool selected, int item, int col, int line, int width, int start = 0)
	{
		if (item >= Count)
		{
			return;
		}

		var str = this[item];

		if (str.Length > width)
		{
			str = str.Substring(0, width);
		}
		else
		{
			str = str.PadRight(width, ' ');
		}

		container.Move(col, line);

		//alternate the colors
		if(item == container.SelectedItem)
        {
			driver.SetAttribute(new Terminal.Gui.Attribute(Color.Cyan, Color.BrightBlue));

			driver.AddStr("OMG YOUR ABOUT TO SELECT IT!");
		}
		else
		{
			if (item % 2 == 0)
			{
				driver.SetAttribute(new Terminal.Gui.Attribute(Color.Green, Color.Blue));
			}
			else
			{
				driver.SetAttribute(new Terminal.Gui.Attribute(Color.Blue, Color.Green));
			}

			driver.AddStr(str);
		}

	}

	public void SetMark(int item, bool value)
	{
		// choosing not to support marking
	}

	public IList ToList()
	{
		return this;
	}
}
```

**Comment by BDisp** [comment]
@tznind I propose that can be added a flag in the `ListView` for to use that feature. You can added it in the `ListWrapper`. What do you think?

**Comment by tznind** [comment]
I don't think alternating colours is something that needs a flag.  Especially since you also have to implicitly consider the selected row as well as when the control is disabled etc.  So really you need to specify a `ColorScheme` not an `Attribute`.  That is how it works for `TableView`:

```csharp
/// <summary>
/// Defines a delegate for returning custom representations of cell values.  If not set then <see cref="object.ToString()"/> is used.  Return values from your delegate may be truncated e.g. based on <see cref="MaxWidth"/>
/// </summary>
public Func<object, string> RepresentationGetter;

/// <summary>
/// Delegate for providing color to <see cref="TableView"/> cells based on the value being rendered
/// </summary>
/// <param name="args">Contains information about the cell for which color is needed</param>
/// <returns></returns>
public delegate ColorScheme CellColorGetterDelegate (CellColorGetterArgs args);

/// <summary>
/// Delegate for providing color for a whole row of a <see cref="TableView"/>
/// </summary>
/// <param name="args"></param>
/// <returns></returns>
public delegate ColorScheme RowColorGetterDelegate (RowColorGetterArgs args);
		
```

For now I think `IListDataSource` is versatile and not too onerous to implement.  A compromise would be looking at `ListViewWrapper` and making it easier to subclass so that users can inherit from `ListViewWrapper` and do the above with less overhead.

**Comment by aurelienverbeke** [comment]
Thank you @tznind. I'll try to use your example and give you feedbacks.

**Comment by BDisp** [comment]
I understand. The `ListViewWrapper` is now public and more easier to inherit, to using this feature

**Comment by aurelienverbeke** [comment]
So following on from @tznind work, I decided to copy the `ListWrapper` class and create the same which includes the color display.

In fact, for each item of the `source` list, there's background and a foreground/text colors associated.

All seem to work for me. (programming is just my hobby so it can certainly be made more efficiently...)

```csharp
public class ColorListWrapper : IListDataSource {
		IList src;
		BitArray marks;
		int count, len;
		Color[] backgroundColors;
		Color[] textColors;

		public ColorListWrapper(IList source, Color[] backgroundColors, Color[] textColors)
		{
			if (source != null)
			{
				count = source.Count;
				marks = new BitArray (count);
				src = source;
				len = GetMaxLengthItem ();
				this.backgroundColors = backgroundColors;
				this.textColors = textColors;
			}
		}

		public int Count => src != null ? src.Count : 0;

		public int Length => len;

		int GetMaxLengthItem ()
		{
			if (src?.Count == 0)
				return 0;

			int maxLength = 0;
			for (int i = 0; i < src.Count; i++)
			{
				var t = src [i];
				int l;
				if (t is ustring u)
					l = u.RuneCount;
				else if (t is string s)
					l = s.Length;
				else
					l = t.ToString ().Length;

				if (l > maxLength)
					maxLength = l;
			}

			return maxLength;
		}

		void RenderUstr (ConsoleDriver driver, ustring ustr, int col, int line, int width, int start = 0)
		{
			int byteLen = ustr.Length;
			int used = 0;
			for (int i = start; i < byteLen;)
			{
				(var rune, var size) = Utf8.DecodeRune (ustr, i, i - byteLen);
				var count = Rune.ColumnWidth (rune);
				if (used + count > width)
					break;
				driver.AddRune (rune);
				used += count;
				i += size;
			}

			for (; used < width; used++)
				driver.AddRune (' ');
		}

		public void Render (ListView container, ConsoleDriver driver, bool marked, int item, int col, int line, int width, int start = 0)
		{
			container.Move (col, line);
			var t = src [item];
			if (t == null)
				RenderUstr (driver, ustring.Make (""), col, line, width);
			else
			{
				if (item != container.SelectedItem)
					driver.SetAttribute(new Terminal.Gui.Attribute(textColors[item], backgroundColors[item]));
				if (t is ustring u)
					RenderUstr (driver, u, col, line, width, start);
				else if (t is string s)
					RenderUstr (driver, s, col, line, width, start);
				else
					RenderUstr (driver, t.ToString (), col, line, width, start);
			}
		}

		public bool IsMarked (int item)
		{
			if (item >= 0 && item < count)
				return marks [item];
			return false;
		}

		public void SetMark (int item, bool value)
		{
			if (item >= 0 && item < count)
				marks [item] = value;
		}

		public IList ToList ()
		{
			return src;
		}
	}
```

**Comment by BDisp** [comment]
That is what I talked before. This feature may be inserted by default in the `ListWrapper`, by adding new parameters to initialize the the background and foreground colors with null values if no colors are provided to not breaking compatibility. If no colors are provided then it will use the default for all items. May be is your first opportunity to submit a pull request to this api.

**Comment by aurelienverbeke** [comment]
In fact, I'm not so familiar with delegates, but I did a pull request #1578 based on what I did for me.
Look at it, and let's see if we can improve the code...

**Comment by BDisp** [comment]
You can use like this:
```cs

			listView = new ListView (new ListWrapper (source)) { Width = Dim.Fill (), Height = Dim.Fill () };

			listView.RowRender += ListView_RowRender;


		private void ListView_RowRender (ListViewRowEventArgs obj)
		{
			if (obj.Row == listView.SelectedItem) {
				return;
			}
			if (obj.Row % 2 == 0) {
				obj.RowAttribute = new Attribute (Color.Green, Color.Magenta);
			} else {
				obj.RowAttribute = new Attribute (Color.Magenta, Color.Green);
			}
		}

```

**Comment by angelobreuer** [comment]
@BDisp  Did you mean @aurelienverbeke ?

**Comment by BDisp** [comment]
> @BDisp Did you mean @aurelienverbeke ?

Yes, sorry. Thanks.

**Comment by BDisp** [comment]
@aurelienverbeke, if you like my solution, please merge it in your pull request.

**Comment by aurelienverbeke** [comment]
@BDisp that's a good idea, I merged it.

**Comment by BDisp** [comment]
@aurelienverbeke you also can change the description of your pr about the way it handle de colors and add "Fixes #1577" to auto close this issue. Thanks.

**Comment by aurelienverbeke** [comment]
All is done !
